flink-clients/src/main/java/org/apache/flink/client/program/rest/RestClusterClient.java

```diff
@@ -371,14 +370,24 @@ public void close() {
                         });
 
         final CompletableFuture<JobSubmitResponseBody> submissionFuture =
-                requestFuture.thenCompose(
-                        requestAndFileUploads ->
-                                sendRetriableRequest(
-                                        JobSubmitHeaders.getInstance(),
-                                        EmptyMessageParameters.getInstance(),
-                                        requestAndFileUploads.f0,
-                                        requestAndFileUploads.f1,
-                                        isConnectionProblemOrServiceUnavailable()));
+                requestFuture
+                        .thenCompose(
+                                requestAndFileUploads ->
+                                        sendRetriableRequest(
+                                                JobSubmitHeaders.getInstance(),
+                                                EmptyMessageParameters.getInstance(),
+                                                requestAndFileUploads.f0,
+                                                requestAndFileUploads.f1,
+                                                isConnectionProblemOrServiceUnavailable()))
+                        .thenApply(
+                                response -> {
```

Comment:
I get why you did it but this seems unnecessarily complicated and not intuitive. We are just abusing the response to return information we already had at the beginning because the current code structure does not readily provide access it.

This also
a) should be logged _before_ the submission, because the operation may take a while or timeout during leader retrieval
b) should especially be logged if the request fails, which we currently don't.

I would suggest to add a `Consumer<URL>` argument to `sendRetriableRequest` that is called on each retry with the `WebMonitorBaseUrl`.

Reply:
Makes sense, I'll try to change this after lunch

Reply:
For a) we only log job name and job id. For b) we now log both successful and failed request alongside the receiving dispatcher.